src/client.test.ts

```diff
@@ -10,26 +11,104 @@ const client = new SkynetClient(portalUrl);
+    // NOTE: the failure tests should be first, so that they run before the
+    // portal URL has been successfully resolved (the client will never try to
+    // resolve again after it resolved successfully).
+
+    it("should throw if portal does not send skynet-portal-api header", async () => {
+      mock.onHead(expectedPortalUrl).replyOnce(200, {}, {});
+
+      // No custom portal passed, so should make request for portal.
+      const client = new SkynetClient();
+
+      await expect(client.portalUrl()).rejects.toThrowError("Could not get portal URL for the given portal");
+    });
+
+    it("should throw if portal does not send headers", async () => {
+      mock.onHead(expectedPortalUrl).replyOnce(200, {});
+
+      // No custom portal passed, so should make request for portal.
+      const client = new SkynetClient();
+
+      await expect(client.portalUrl()).rejects.toThrowError(
+        "Did not get 'headers' in response despite a successful request. Please try again and report this issue to the devs if it persists."
+      );
+    });
+
+    // Success cases.
+
+    it("should use default portal if portal URL is not given", async () => {
+      mock.onHead(expectedPortalUrl).replyOnce(200, {}, { "skynet-portal-api": expectedPortalUrl });
+
+      const client = new SkynetClient();
+
+      const portalUrl = await client.portalUrl();
+
+      expect(portalUrl).toEqual(expectedPortalUrl);
+
+      expect(mock.history.get.length).toBe(0);
+    });
+  });
 });
 
 describe("buildRequestUrl", () => {
-  it("should build a URL from the given components", async () => {
-    const endpointPath = "/skynet/foo";
-    const extraPath = "bar";
-    const query = { foo: "bar" };
-    const expectedUrl = `${portalUrl}/skynet/foo/bar?foo=bar`;
+  const endpointPath = "/skynet/foo";
+  const subdomain = "account";
+  const extraPath = "bar";
+  const query = { foo: "bar" };
+
+  it("should build a URL from the given components, using the override URL", async () => {
+    const overrideUrl = "siasky.dev";
+    const expectedUrl = `https://account.${overrideUrl}/skynet/foo/bar?foo=bar`;
 
-    const url = await buildRequestUrl(client, endpointPath, undefined, extraPath, query);
+    const url = await buildRequestUrl(client, endpointPath, overrideUrl, subdomain, extraPath, query);
     expect(url).toEqual(expectedUrl);
   });
+
+  it("should build a URL from the given components, using the portal URL", async () => {
+    const expectedUrl = `https://account.siasky.net/skynet/foo/bar?foo=bar`;
+
+    const url = await buildRequestUrl(client, endpointPath, undefined, subdomain, extraPath, query);
+    expect(url).toEqual(expectedUrl);
+  });
+
+  describe("localhost inputs", () => {
+    const validExpectedLocalhosts = combineStrings(
+      ["https://", "https:", "http://", "http:"],
+      ["localhost"],
+      ["", "/"]
+    );
+    const localhostUrls = combineStrings(["", "https://", "https:", "http://", "http:"], ["localhost"], ["", "/"]);
```

Comment:
`localhostUrls` contains `localhost` and `localhost/` but `validExpectedLocalhosts` does not. Is that intentional? I tried to drill down and see if anything is making sure we always have a protocol string but got a little stuck with `url-parse.parse()`.

Reply:
Yeah. I'll add some comments for clarity.